src/Storages/StorageS3.cpp

```diff
@@ -1754,6 +1983,16 @@ namespace
                     prev_read_keys_size = read_keys.size();
                 }
 
+                if (!format)
+                {
+                    if (current_key_with_info->path_in_archive.has_value())
```

Comment:
AFAIS this code is not needed if we processed `read_keys` correctly. Now they do not seem to contain `path_in_archive` because they are filled from `basic_iterator` instead of `archive_iterator`, which will have actual file names of archive files. Or at least we should not execute the code above related to `read_keys` as it does not make sense for archives.

Reply:
Also I see that format detection doesn't work and schema inference cache is not used correctly when we read from archives:
```
avogar-dev :) desc s3('http://localhost:11111/test/archive.zip::data')

DESCRIBE TABLE s3('http://localhost:11111/test/archive.zip::data')

Query id: ca96b8cc-c0ce-442b-be7d-0cfc7405d3d3


Elapsed: 0.038 sec.

Received exception:
Code: 715. DB::Exception: The data format cannot be detected by the contents of the files. You can specify the format manually. (CANNOT_DETECT_FORMAT)

avogar-dev :) desc s3('http://localhost:11111/test/archive.zip::data', JSONEachRow)

DESCRIBE TABLE s3('http://localhost:11111/test/archive.zip::data', JSONEachRow)

Query id: 2f676400-644c-4976-81a2-7703259e0a64

   ┌─name─┬─type────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
1. │ a    │ Nullable(Int64) │              │                    │         │                  │                │
   └──────┴─────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

1 row in set. Elapsed: 0.008 sec.

avogar-dev :) select count() from s3('http://localhost:11111/test/archive.zip::data', JSONEachRow)

SELECT count()
FROM s3('http://localhost:11111/test/archive.zip::data', JSONEachRow)

Query id: cde8da08-2a5e-4f03-9a79-b397baeeba5f

   ┌─count()─┐
1. │       1 │
   └─────────┘

1 row in set. Elapsed: 0.013 sec.

avogar-dev :) select * from system.schema_inference_cache \G

SELECT *
FROM system.schema_inference_cache

Query id: 8ba2e455-4c1c-46c7-b135-ea79f5dfa84e

Row 1:
──────
storage:                S3
source:                 localhost11111/test/archive.zip
format:                 JSONEachRow
additional_format_info: schema_inference_hints=, max_rows_to_read_for_schema_inference=25000, max_bytes_to_read_for_schema_inference=25000, schema_inference_make_columns_nullable=true, try_infer_integers=true, try_infer_dates=true, try_infer_datetimes=true, try_infer_numbers_from_strings=false, read_bools_as_numbers=true, read_bools_as_strings=true, read_objects_as_strings=true, read_numbers_as_strings=true, read_arrays_as_strings=true, try_infer_objects_as_tuples=true, infer_incomplete_types_as_strings=true, try_infer_objects=false, use_string_type_for_ambiguous_paths_in_named_tuples_inference_from_objects=false
registration_time:      2024-04-24 13:19:26
schema:                 a Nullable(Int64)
number_of_rows:         ᴺᵁᴸᴸ
schema_inference_mode:  DEFAULT

Row 2:
──────
storage:                S3
source:                 localhost11111/test/data
format:                 JSONEachRow
additional_format_info: schema_inference_hints=, max_rows_to_read_for_schema_inference=25000, max_bytes_to_read_for_schema_inference=25000, schema_inference_make_columns_nullable=true, try_infer_integers=true, try_infer_dates=true, try_infer_datetimes=true, try_infer_numbers_from_strings=false, read_bools_as_numbers=true, read_bools_as_strings=true, read_objects_as_strings=true, read_numbers_as_strings=true, read_arrays_as_strings=true, try_infer_objects_as_tuples=true, infer_incomplete_types_as_strings=true, try_infer_objects=false, use_string_type_for_ambiguous_paths_in_named_tuples_inference_from_objects=false
registration_time:      2024-04-24 13:19:30
schema:                 ᴺᵁᴸᴸ
number_of_rows:         1
schema_inference_mode:  DEFAULT

2 rows in set. Elapsed: 0.001 sec.
```

In this case schema inference cache should contain single entry with path `localhost11111/test/archive.zip::data`. There should be tests for it (we have similar tests for reading from archives in `File` engine)